### PR TITLE
Improve Type Hinting

### PR DESCRIPTION
**Pull request by rdohms**
- [x] Install PHPStan with level 8 and plugins
- [x] Address initial issues in `src` and `tests`

Ok, this solves all main type issues and a couple of small issues here and there that PHPstorm was eager to fix also.

Let's look it over carefully, but all tests are good. I saw some other small issues along the way, I'll try to give some time to that also, and make a few suggestions to tighten everything together.

I updated the actions to separate checks from tests to get contributors early results.

### CODE REVIEW

**Comment by rdohms** [comment]
@LKaemmerling it seems i do not have any permissions on this repo still even after being added to the org. Can you give me admin? want to tweak some branch protections and such.

File: src/Prometheus/Histogram.php [hunk 1]
```diff
@@ -12,27 +12,33 @@ class Histogram extends Collector
...
-     * @param string $name
-     * @param string $help
-     * @param array $labels
-     * @param array $buckets
+     * @param Adapter    $adapter
+     * @param string     $namespace
+     * @param string     $name
+     * @param string     $help
+     * @param string[]   $labels
+     * @param mixed[]|null $buckets
```
Comment (LKaemmerling) [inline src/Prometheus/Histogram.php:25]: I'm not quite sure if `mixed` is the correct type here. 

I looked into the official go client and there it is always a float: https://github.com/prometheus/client_golang/blob/47cfdc9bb8ceaf18b12a4b7ba778b06e12d1a97a/prometheus/histogram.go#L63

So maybe we should also use a float here?

File: src/Prometheus/Histogram.php [hunk 2]
```diff
@@ -52,20 +58,35 @@ public function __construct(Adapter $adapter, $namespace, $name, $help, $labels
 
     /**
      * List of default buckets suitable for typical web application latency metrics
-     * @return array
+     *
+     * @return mixed[]
```
Comment (LKaemmerling) [inline src/Prometheus/Histogram.php:62]:
```suggestion
     * @return float[]

File: src/Prometheus/Histogram.php [hunk 3]
```diff
@@ -52,20 +58,35 @@ public function __construct(Adapter $adapter, $namespace, $name, $help, $labels
...
     }
 
     /**
      * @param float $start
      * @param float $growthFactor
-     * @param int $numberOfBuckets
-     * @return array
+     * @param int   $numberOfBuckets
+     *
+     * @return mixed[]
```
Comment (LKaemmerling) [inline src/Prometheus/Histogram.php:89]:
```suggestion
     * @return float[]

File: src/Prometheus/RegistryInterface.php [hunk 4]
```diff
@@ -61,30 +61,36 @@ public function registerCounter($namespace, $name, $help, $labels = []): Counter
...
-     * @param string $namespace e.g. cms
-     * @param string $name e.g. duration_seconds
-     * @param string $help e.g. A histogram of the duration in seconds.
-     * @param array  $labels e.g. ['controller', 'action']
-     * @param array  $buckets e.g. [100, 200, 300]
+     * @param string   $namespace e.g. cms
+     * @param string   $name e.g. duration_seconds
+     * @param string   $help e.g. A histogram of the duration in seconds.
+     * @param string[] $labels e.g. ['controller', 'action']
+     * @param int[]    $buckets e.g. [100, 200, 300]
```
Comment (LKaemmerling) [inline src/Prometheus/RegistryInterface.php:82]: Should be floats too

File: src/Prometheus/RegistryInterface.php [hunk 5]
```diff
@@ -93,17 +99,17 @@ public function registerHistogram($namespace, $name, $help, $labels = [], $bucke
...
+    public function getHistogram(string $namespace, string $name): Histogram;
 
     /**
      * @param string $namespace e.g. cms
      * @param string $name e.g. duration_seconds
      * @param string $help e.g. A histogram of the duration in seconds.
-     * @param array  $labels e.g. ['controller', 'action']
-     * @param array  $buckets e.g. [100, 200, 300]
+     * @param string[]  $labels e.g. ['controller', 'action']
+     * @param int[]  $buckets e.g. [100, 200, 300]
```
Comment (LKaemmerling) [inline src/Prometheus/RegistryInterface.php:109]: Floats too

File: src/Prometheus/RenderTextFormat.php [hunk 6]
```diff
@@ -36,14 +36,9 @@ public function render(array $metrics): string
...
     {
-        $escapedLabels = [];
-
         $labelNames = $metric->getLabelNames();
         if ($metric->hasLabelNames() || $sample->hasLabelNames()) {
-            $labels = array_combine(array_merge($labelNames, $sample->getLabelNames()), $sample->getLabelValues());
-            foreach ($labels as $labelName => $labelValue) {
-                $escapedLabels[] = $labelName . '="' . $this->escapeLabelValue($labelValue) . '"';
-            }
+            $escapedLabels = $this->escapeAllLabels($labelNames, $sample);
```
Comment (LKaemmerling) [inline src/Prometheus/RenderTextFormat.php:41]: Good refactor 👍  it is now more readable!

File: src/Prometheus/RenderTextFormat.php [hunk 7]
```diff
@@ -53,11 +48,31 @@ private function renderSample(MetricFamilySamples $metric, Sample $sample): stri
...
-        $v = str_replace("\\", "\\\\", $v);
-        $v = str_replace("\n", "\\n", $v);
-        $v = str_replace("\"", "\\\"", $v);
-        return $v;
+        $escapedLabels = [];
+
+        $labels = array_combine(array_merge($labelNames, $sample->getLabelNames()), $sample->getLabelValues());
+
+        if ($labels === false) {
+            return $escapedLabels;
```
Comment (LKaemmerling) [inline src/Prometheus/RenderTextFormat.php:69]: I would prefer to return an empty array here, because it is easier to reed.
```suggestion
            return [];

File: src/Prometheus/Storage/APC.php [hunk 8]
```diff
@@ -170,6 +171,7 @@ private function collectCounters(): array
                 'help' => $metaData['help'],
                 'type' => $metaData['type'],
                 'labelNames' => $metaData['labelNames'],
+                'samples' => [],
```
Comment (LKaemmerling) [inline src/Prometheus/Storage/APC.php:174]: I'm not quite sure if this won't change the printed/restored result, can you check this?

File: src/Prometheus/Storage/APC.php [hunk 9]
```diff
@@ -200,6 +202,7 @@ private function collectGauges(): array
...
```
Comment (LKaemmerling) [inline src/Prometheus/Storage/APC.php:205]: Same as above

File: src/Prometheus/Storage/InMemory.php [hunk 10]
```diff
@@ -10,8 +10,19 @@
 class InMemory implements Adapter
 {
 
+    /**
+     * @var mixed[]
+     */
     private $counters = [];
```
Comment (LKaemmerling) [inline src/Prometheus/Storage/InMemory.php:16]: Should be a Counter[]

File: src/Prometheus/Storage/InMemory.php [hunk 11]
```diff
@@ -10,8 +10,19 @@
...
+
+    /**
+     * @var mixed[]
```
Comment (LKaemmerling) [inline src/Prometheus/Storage/InMemory.php:19]: Gauge[]

File: src/Prometheus/Storage/InMemory.php [hunk 12]
```diff
@@ -10,8 +10,19 @@
...
+     */
     private $gauges = [];
+
+    /**
+     * @var mixed[]
```
Comment (LKaemmerling) [inline src/Prometheus/Storage/InMemory.php:24]: Histogram[]

File: tests/Test/BlackBoxTest.php [hunk 13]
```diff
@@ -45,20 +52,20 @@ public function gaugesShouldBeOverwritten()
         $metricsResult = $this->client->get('/examples/metrics.php?adapter=' . $this->adapter);
         $body = (string)$metricsResult->getBody();
         echo "\nbody: " . $body . "\n";
-        $this->assertThat(
+        self::assertThat(
```
Comment (LKaemmerling) [inline tests/Test/BlackBoxTest.php:55]: Why the change from `$this->` to `self::`?

**Review by LKaemmerling — CHANGES_REQUESTED** [review]
Thank you, this is looking really good. I have a few smaller comments, for most of the `mixed` types we should be able to use already existing objects as type hints.

**Comment by LKaemmerling** [comment]
> @LKaemmerling it seems i do not have any permissions on this repo still even after being added to the org. Can you give me admin? want to tweak some branch protections and such.

@rdohms I changed the Branch Protection Policy a bit. If you have some other ideas we should talk about it :)

**Comment by rdohms** [comment]
@LKaemmerling i'll fidget with the mixed values, but it may require restricting more places that now are flexible. I'll see what I can further restrict.

Regarding permissions, I also can't seem to create branches, assign labels or anything. Just wondering which role you would like me to play in the org.

File: src/Prometheus/RenderTextFormat.php (hunk 7 above)
Comment (rdohms) [reply src/Prometheus/RenderTextFormat.php:69]: I had this originally as well, agree its more readable

File: src/Prometheus/Storage/APC.php (hunk 8 above)
Comment (rdohms) [reply src/Prometheus/Storage/APC.php:174]: I'll do some tests, but the code is wonky and I'm not sure where it comes from and where its going, I'll dig a bit more.

File: tests/Test/BlackBoxTest.php (hunk 13 above)
Comment (rdohms) [reply tests/Test/BlackBoxTest.php:55]: so, while Sebastian has no preference, the functions are all static, so PHPStan will complain about an instance invocation instead of a static invocation. For these reasons most projects I have worked on resort to using `self::`. So I just instinctively changed it, do you have a preference for this?

File: src/Prometheus/Storage/InMemory.php (hunk 10 above)
Comment (rdohms) [reply src/Prometheus/Storage/InMemory.php:16]: So according to lines 221-224 this is not actually the case:
```
            $this->counters[$metaKey] = [
                'meta' => $this->metaData($data),
                'samples' => [],
            ];
```

This is the same for gauges and histograms. The Counter/Gauge/Histogram classes are not even imported in the `use` section.

File: src/Prometheus/Storage/InMemory.php (hunk 11 above)
Comment (rdohms) [reply src/Prometheus/Storage/InMemory.php:19]: see above in `$counters`

File: src/Prometheus/Storage/InMemory.php (hunk 12 above)
Comment (rdohms) [reply src/Prometheus/Storage/InMemory.php:24]: see above in `$counters`

File: src/Prometheus/Storage/APC.php (hunk 8 above)
Comment (rdohms) [reply src/Prometheus/Storage/APC.php:174]: Got the blackbox testing up and running again, and it looks all good.

```
# HELP php_info Information about the PHP environment.
# TYPE php_info gauge
php_info{version="7.4.11"} 1
# HELP test_some_gauge it sets
# TYPE test_some_gauge gauge
test_some_gauge{type="blue"} 0
```

But I noticed 2 things I'll address:
- there are no tests for the renderer
- blackbox was outdated and not in the GHA workflow

Comment (rdohms) [reply src/Prometheus/Storage/APC.php:174]: After further testing, I noticed that this behaviour of an empty `samples` array is also done in the `MetricsFamilySamples`, so doing it here will have no effect that won't already happen there.

**Comment by rdohms** [comment]
Alright, comments addressed.

File: src/Prometheus/Storage/APC.php (hunk 8 above)
Comment (LKaemmerling) [reply src/Prometheus/Storage/APC.php:174]: Okay thank you for digging more into it 👍

File: src/Prometheus/Storage/InMemory.php (hunk 10 above)
Comment (LKaemmerling) [reply src/Prometheus/Storage/InMemory.php:16]: mmh yep you are right. I personally think in the future we should cleanup this mix of using arrays within arrays on the one side and arrays using objects on the other side.

File: tests/Test/BlackBoxTest.php (hunk 13 above)
Comment (LKaemmerling) [reply tests/Test/BlackBoxTest.php:55]: Nope i'm fine with the static approach too, even if it is a bit strange in the first place.

**Review by LKaemmerling — APPROVED** [review]
Good job, thank you! I will test this branch in my test & production environment and if everything is still fine i will merge it. Thank you again!

**Comment by LKaemmerling** [comment]
@rdohms i tested it a bit on my test instance (1k metrics, scraped every 10sec) and it works really great! Great job
